src/react/types.ts

```diff
@@ -18,8 +20,16 @@ export interface UseBaseQueryOptions<
 export interface UseQueryOptions<
   TQueryFnData = unknown,
   TError = unknown,
-  TData = TQueryFnData
-> extends UseBaseQueryOptions<TQueryFnData, TError, TData> {}
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey
+>
+  extends Omit<
+    UseBaseQueryOptions<TQueryFnData, TError, TData>,
+    'queryFn' | 'queryKey'
+  > {
```

Comment:
why not change this on `UseBaseQueryOptions` directly? As far as I can see, this is the only interface that extends `UseBaseQueryOptions`....

Reply:
there is a function in `src/react/useBaseQuery.ts:9` that uses that uses the UseBaseQueryOptions interface, I mainly did it this way to keep the impact minimal

Reply:
I think this goes hand-in-hand with the request for infinite query support, because `src/react/useBaseQuery.ts:9` is only used for `useQuery` - which you have already adapted, and `useInfiniteQuery`. With regards to the queryKey, they are identical, so if you add useInfiniteQuery support to this PR, you could also get rid of the Omit I think.

Reply:
Yes I see what you mean now. I will update it accordingly.

Reply:
I had a check and I don't think I can make it work like that the interfaces are put together like this

```
interface UseQueryOptions extends UseBaseQueryOptions extends QueryObserverOptions {}
interface UseInfiniteQueryOptions extends InfiniteQueryObserverOptions extends QueryObserverOptions {}
```

While it is true that useQuery and useInfiniteQuery do both use the useBaseQuery function, I don't think that matters when it comes to inferring the queryKey (which is either inferred through `QueryFunction` or `UseQueryOptions`).

Reply:
@boschni what do you think?

Reply:
Think we should try to keep the API consistent and implement things like this library wide instead of in one hook

Reply:
@boschni This pr includes no API changes, only a minor change to an interface to resolve a minor TypeScript inconvenience. I am not saying the interfaces for ouseInfiniteQuery can't be changed in order for inference to work. I'm saying it can't be done as @TkDodo [describes](https://github.com/tannerlinsley/react-query/pull/1955#discussion_r593250990)

Reply:
If someone is opting in to use this feature, they most likely want to use it throughout their codebase, and not only when using `useQuery`. At this moment you might only use `useQuery`, but others are also using `queryClient.fetchQuery` or `useInfiniteQuery`. If we provide support for this pattern, users should be able to expect that they can use it anywhere.

Reply:
I’ll try to take another stab at this if i find the time - i think we’re already on the right track here 👍